### PR TITLE
fix: Change is_last chunk calc and add conditional break in chunk_iter

### LINKED ISSUE #21568: Incorrect right stride when sequence end falls within right stride of penultimate chunk
### System Info

- `transformers` version: 4.26.0
- Platform: Linux-4.14.301-224.520.amzn2.x86_64-x86_64-with-glibc2.26
- Python version: 3.9.15
- Huggingface_hub version: 0.12.0
- PyTorch version (GPU?): 1.11.0 (False)
- Tensorflow version (GPU?): 2.7.1 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker @sanchit-gandhi @Narsil

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The easiest way to see this problem is to use a Whisper ASR model and debug into the `chunk_iter` method of `transformers/src/transformers/pipelines/automatic_speech_recognition.py` with an audio file that is of a specific length. An example length that illustrates this problem is about 87 seconds in length. The specific file I was testing with had 1,385,995 samples at 16kHz.

### Expected behavior

The Whisper model requires chunks of 30 seconds in length. I tested using the default stride of 5 seconds on each side. So a chunk is `30*16,000=480,000` samples. The stride width is normally `5*16,000=80,000 `samples. But the first chunk doesn't have a left stride and the final chunk doesn't have a right stride so, for this file we would expect the following chunk sizes and strides...

* 480,000; 0; 80,000
* 480,000; 80,000; 80,000
* 480,000; 80,000; 80,000
* 425,995; 80,000; **25,995**
* 105,995; 80,000; 0

But what we actually get is

* 480,000; 0; 80,000
* 480,000; 80,000; 80,000
* 480,000; 80,000; 80,000
* 425,995; 80,000; **80,000**
* 105,995; 80,000; 0

This means that the fourth (penultimate) chunk is processed as if it has a full right stride when in fact it shouldn't.

### PR DESCRIPTION
**Pull request by connor-henderson**
# What does this PR do?
Fixes #21568 

With three functional changes
- Updates the is_last calc to include the right_stride
- Conditionally breaks at the end of the loop block if is_last is true
- Adds to a test

And two non-functional changes
- Renamed `i` to `chunk_start_idx` and put `chunk_end_idx` in a variable
- Removed a comment and added another

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
[Issue](https://github.com/huggingface/transformers/issues/21568)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@Narsil

### CODE REVIEW

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 1]
```diff
@@ -1109,6 +1109,11 @@ def test_chunk_iterator_stride(self):
         self.assertEqual(len(outs), 2)
         self.assertEqual([o["stride"] for o in outs], [(90, 0, 0), (30, 20, 0)])
         self.assertEqual([o["input_values"].shape for o in outs], [(1, 90), (1, 30)])
+        
+        outs = list(chunk_iter(inputs, feature_extractor, 36, 6, 6, ratio))
```
Comment (connor-henderson) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:1113]: Roughly scaled down params from the example in the [issue](https://github.com/huggingface/transformers/issues/21568)

**Comment by Narsil** [comment]
@ArthurZucker Could you please review this ?

I am ok with this PR, but since I made the proposed change, I'm too heavily biased to review properly.

I will take another look still now that I might be able to think straighter now.

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 2]
```diff
@@ -56,14 +56,15 @@ def rescale_stride(stride, ratio):
...
+        chunk_end_idx = chunk_start_idx + chunk_len
+        chunk = inputs[chunk_start_idx:chunk_end_idx]
         processed = feature_extractor(chunk, sampling_rate=feature_extractor.sampling_rate, return_tensors="pt")
         if dtype is not None:
             processed = processed.to(dtype=dtype)
-        _stride_left = 0 if i == 0 else stride_left
-        is_last = i + step + stride_left >= inputs_len
+        _stride_left = 0 if chunk_start_idx == 0 else stride_left
+        # all right strides must be full, otherwise it is the last item
+        is_last = chunk_end_idx > inputs_len if stride_right > 0 else chunk_end_idx >= inputs_len
```
Comment (ArthurZucker) [inline src/transformers/pipelines/automatic_speech_recognition.py:67]: this is the "biggest" change, and yeah, makes sense.

**Review by ArthurZucker — APPROVED** [review]
Looks good to me! Thanks for this contribution!

**Comment by connor-henderson** [comment]
> Just thinking we should update the values of the whisper timestamps (so try running the slow tests for the ASR pipeline, only those related to whisper)

Updated the timestamps to update after running the below slow asr pipeline tests:
`test_return_timestamps_in_preprocess`
`test_torch_whisper`
`test_find_longest_common_subsequence`
`test_whisper_timestamp_prediction`
`test_simple_whisper_asr`
`test_simple_whisper_translation`

**Comment by Narsil** [comment]
Thank you very much for this ! And detecting this bug !

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 3]
```diff
@@ -548,11 +548,11 @@ def test_whisper_timestamp_prediction(self):
                     },
                     {
                         "text": " the thousands of spectators, retrievality is not worth thinking about.",
-                        "timestamp": (19.6, 24.98),
+                        "timestamp": (19.6, 26.66),
                     },
                     {
                         "text": " His instant panic was followed by a small, sharp blow high on his chest.",
-                        "timestamp": (24.98, 30.98),
+                        "timestamp": (26.66, 31.060000000000002),
```
Comment (sgugger) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:555]: Can we use something that tests the almost equality and replace with this 31.06? You can use `nested_simplify` on the output at line 529 and this should work.

```suggestion
                        "timestamp": (26.66, 31.06),
```

**Review by sgugger — APPROVED** [review]
Thanks for the fix! Just one small comment on the test.

**Comment by sgugger** [comment]
Thanks again for your contribution!

**Pull request merged**